### PR TITLE
citra: rename to citra-nightly, init citra-canary

### PR DESCRIPTION
**Pull request by kira64xyz**
This updates the citra derivation to a stable release and also adds support for another branch, citra-nightly.
See the commit description for details.

This also enables fetching the in-app compatibility list.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/emulators/citra/generic.nix [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+    # Devendor some submodules
+    rm -rf externals/{libressl,zstd}
+    tar -xf ${libressl.src} --strip-components 1 --one-top-level="externals/libressl"
+    cp -r ${zstd.src} ./externals/zstd
+  '';
+
+  preConfigure = ''
+    # Trick the configure system. CMakeLists.txt#L55 function(check_submodules_present)
+    chmod 775 -R ./externals
```
Comment (IvarWithoutBones) [inline pkgs/applications/emulators/citra/generic.nix:95]: I think yuzu has a way nicer fix for this, we should make it possible to do the same here and upstream that some day.

**Review by IvarWithoutBones — COMMENTED** [review]
Thanks for the work! It's much appreciated :blush:

File: pkgs/applications/emulators/citra/generic.nix [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    "-DDYNARMIC_ENABLE_CPU_FEATURE_DETECTION=OFF"
+    "-DENABLE_COMPATIBILITY_LIST_DOWNLOAD=OFF" # Sacrifices reproducibility
+    "-DENABLE_FFMPEG_VIDEO_DUMPER=ON"
+    "-DENABLE_QT_TRANSLATION=ON"
+    "-DUSE_SYSTEM_BOOST=ON"
+    "-GNinja"
+    "-Wno-dev"
+    "-DUSE_DISCORD_PRESENCE=${if discordSupport then "ON" else "OFF"}"
+    "-DENABLE_WEB_SERVICE=${if onlineSupport then "ON" else "OFF"}"
+    (if fdkSupport then "-DENABLE_FDK=ON" else "-DENABLE_FFMPEG_AUDIO_DECODER=ON")
```
[the remark this answers is not in the document]
Comment (SuperSamus) [reply pkgs/applications/emulators/citra/generic.nix:55]: Yes, if both are enabled FDK is silently disabled.
(Though both being disabled could be an option.)

File: pkgs/applications/emulators/citra/generic.nix [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+stdenv.mkDerivation rec {
+  inherit pname version src;
+
+  cmakeFlags = [
+    "-DBUILD_SHARED_LIBS=OFF"
+    "-DCITRA_ENABLE_COMPATIBILITY_REPORTING=ON"
+    "-DCITRA_USE_BUNDLED_FFMPEG=OFF"
+    "-DCITRA_USE_BUNDLED_QT=OFF"
+    "-DCITRA_USE_BUNDLED_SDL2=OFF"
+    "-DDYNARMIC_ENABLE_CPU_FEATURE_DETECTION=OFF"
```
Comment (SuperSamus) [inline pkgs/applications/emulators/citra/generic.nix:46]: Is this necessary?
I gave a quick look at Dynarmic's source code, [it seems like](https://github.com/merryhime/dynarmic/blob/644172477eaf0d822178cb7e96c62b75caa96573/src/dynarmic/backend/x64/block_of_code.cpp#L78) CPU feature detection is done at runtime, not compile time.

**Comment by kira64xyz** [comment]
Thank you for your feedback everyone! It appears I had based my changes on an outdated version of the derivation, however everything should be good now :blush:

File: pkgs/applications/emulators/citra/generic.nix (hunk 3 above)
Comment (kira64xyz) [reply pkgs/applications/emulators/citra/generic.nix]: Right, this was a remnant of an old version of the derivation. Has been removed, thanks :blush:

File: pkgs/applications/emulators/citra/generic.nix [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    rm -rf externals/zstd externals/libressl
+    cp -r ${zstd.src} externals/zstd
+    tar xf ${libressl.src} -C externals/
+    mv externals/${libressl.name} externals/libressl
+    chmod -R a+w externals/zstd
+  '';
+
+  # Fixes https://github.com/NixOS/nixpkgs/issues/171173
+  postInstall = lib.optionalString (enableCubeb && enableSdl2) ''
+    wrapProgram "$out/bin/citra" \
```
Comment (IvarWithoutBones) [inline pkgs/applications/emulators/citra/generic.nix:85]: Should this be applied to `citra-qt` as well?

File: pkgs/applications/emulators/citra/update.sh [hunk 5]
```diff
@@ -0,0 +1,84 @@
+#! /usr/bin/env nix-shell
+#! nix-shell -i bash -p nix nix-prefetch-git coreutils curl jq gnused
+
+set -e
```
Comment (IvarWithoutBones) [inline pkgs/applications/emulators/citra/update.sh:4]: Can we `set -eou pipefail`?

File: pkgs/applications/emulators/citra/generic.nix (hunk 4 above)
Comment (SuperSamus) [reply pkgs/applications/emulators/citra/generic.nix:85]: The issue linked only affects the SDL interface for some reason, so it's not necessary for `citra-qt`.
(Also this is from #171955)

Comment (kira64xyz) [reply pkgs/applications/emulators/citra/generic.nix:85]: The linked issue implies the Qt backend is not affected by this issue, so I don't think so

File: pkgs/applications/emulators/citra/update.sh (hunk 5 above)
Comment (kira64xyz) [reply pkgs/applications/emulators/citra/update.sh]: Added, thanks :blush:

**Review by IvarWithoutBones — APPROVED** [review]
Looks good to me, thank's a lot for the work! Very happy to see the citra derivation being up to speed with yuzu's 😊

Unless anyone has any issues with it, I'll go ahead and merge this in the coming few days 👍

**Pull request merged**